File: Trees (BT + BST)/cpp/diameterOfTree.cpp
// Problem Statement: Find the Diameter of a Binary Tree. Diameter is the length of the longest path between any 2 nodes in the tree and this path may or may not pass from the root.

// Problem Link: https://leetcode.com/problems/diameter-of-binary-tree/

#include <iostream>
#include <algorithm>
using namespace std;

class TreeNode {
public:
    int val;
    TreeNode* left;
    TreeNode* right;

    TreeNode() {
        this->val = 0;
        this->left = NULL;
        this->right = NULL;
    }

    TreeNode(int val) {
        this->val = val;
        this->left = NULL;
        this->right = NULL;
    }

    TreeNode(int val, TreeNode* left, TreeNode* right) {
        this->val = val;
        this->left = left;
        this->right = right;
    }
};

class Solution {
private:
    int height(TreeNode* root, int& diameter) {
        if (root == NULL) {
            return 0;
        }

        int leftHeight = height(root->left, diameter);
        int rightHeight = height(root->right, diameter);

        diameter = max(diameter, leftHeight + rightHeight);

        return 1 + max(leftHeight, rightHeight);
    }

public:
    int diameterOfBinaryTree(TreeNode* root) {
        int diameter = 0;
        height(root, diameter);
        return diameter;
    }
};

File: Trees (BT + BST)/python/diameterOfTree.py
# Problem Statement: Find the Diameter of a Binary Tree. Diameter is the length of the longest path between any 2 nodes in the tree and this path may or may not pass from the root.

# Problem Link: https://leetcode.com/problems/diameter-of-binary-tree/

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def diameterOfBinaryTree(self, root):
        self.diameter = 0
        self.height(root)
        return self.diameter
        
    def height(self, root):
        if not root:
            return 0
            
        left_height = self.height(root.left)
        right_height = self.height(root.right)
        
        self.diameter = max(self.diameter, left_height + right_height)
        
        return 1 + max(left_height, right_height)
